gencode.c

```diff
@@ -8175,6 +8175,53 @@ gen_multicast(compiler_state_t *cstate, int proto)
 	/*NOTREACHED*/
 }
 
+struct block *
+gen_ifindex(compiler_state_t *cstate, int ifindex)
+{
+	register struct block *b0;
+
+	/*
+	 * Catch errors reported by us and routines below us, and return NULL
+	 * on an error.
+	 */
+	if (setjmp(cstate->top_ctx))
+		return (NULL);
+
+	/*
+	 * Only some data link types support ifindex qualifiers.
+	 */
+	switch (cstate->linktype) {
+	case DLT_LINUX_SLL2:
+		/* match packets on this interface */
+		b0 = gen_cmp(cstate, OR_LINKHDR, 4, BPF_W, ifindex);
```

Comment:
DLT_LINUX_SLL and DLT_LINUX_SLL2 headers are generated by libpcap; they're not visible to the in-kernel filter engine.

This probably needs to be done with something such as

    if (doing a live capture) {
        #ifdef linux
            test SKF_AD_OFF + SKF_AD_IFINDEX
        #else
            report "only supported on Linux"
        #endif
    } else {
        if (linktype is DLT_LINUX_SLL2)
            test the field at an offset of 4
        else
            report "not supported on savefiles except for DLT_LINUX_SLL2
    }

Reply:
> DLT_LINUX_SLL and DLT_LINUX_SLL2 headers are generated by libpcap; they're not visible to the in-kernel filter engine.

Right, that's why `pcap-linux.c:fix_offset()` changes the filter on the fly when it's `DLT_LINUX_SLL*`.

> This probably needs to be done with something such as

I think my code is equivalent to what you suggest, but has the advantage of having strong parallels with the existing `gen_inbound()`.  Note that if we're doing a live capture, the code in `pcap-linux.c:fix_offset()` will turn the offset 4 comparison into an offset `SKF_AD_OFF + SKF_AD_IFINDEX` comparison.

Reply:
OK.